### PR TITLE
Allow position_ids_start=2 on DataCollatorWithFlattening for RoBERTa etc.

### LINKED ISSUE #47496: `DataCollatorWithFlattening` uses 0-based `position_ids`, silently degrading RoBERTa-family models
### System Info

Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

Copy-and-paste the text below in your GitHub issue and FILL OUT the two last points.

- `transformers` version: 5.14.1
- Platform: Linux-6.18.33.2-microsoft-standard-WSL2-x86_64-with-glibc2.35
- Python version: 3.11.13
- Huggingface_hub version: 1.24.0
- Safetensors version: 0.8.0
- Accelerate version: 1.5.2
- Accelerate config:    not found
- DeepSpeed version: 0.17.6
- PyTorch version (accelerator?): 2.11.0+cu128 (CUDA)
- Using distributed or parallel set-up in script?: no
- Using GPU in script?: yes
- GPU type: NVIDIA GeForce RTX 3090

### Who can help?

@Cyrilvallez @vasqu

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Simply put, the `DataCollatorWithFlattening` creates `position_ids` using just `list(range(len(input_ids))`. Works well for the vast majority of models, but there's a handful of architectures (e.g. RoBERTa) that start their position_ids at 2 (or `padding_idx + 1` to be specific).

This means that using DataCollatorWithFlattening works with the vast majority of models, but breaks for a handful of architectures.

```python
import torch
from transformers import AutoModel, AutoTokenizer, DataCollatorWithFlattening

model_id = "FacebookAI/xlm-roberta-base"
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModel.from_pretrained(model_id, attn_implementation="kernels-community/flash-attn2@v3", device_map="auto", torch_dtype="float16")

# 1. Standard tokenization with padding
standard_tokenized = tokenizer(["Hello world", "A second, slightly longer example sentence"], return_tensors="pt", padding=True)
standard_tokenized.to(model.device)

# 2. Tokenization with flattening instead
features = [tokenizer(text) for text in ["Hello world", "A second, slightly longer example sentence"]]
collator = DataCollatorWithFlattening(return_position_ids=True)
flattened_tokenized = collator(features, return_tensors="pt")
flattened_tokenized = {k: v.to(model.device) for k, v in flattened_tokenized.items()}
print(flattened_tokenized["position_ids"])
# tensor([[0, 1, 2, 3, 0, 1, 2, 3, 4, 5, 6, 7, 8]])
# XLM-R computes [2, 3, 4, 5] for the first sequence when no position_ids are passed.

with torch.no_grad():
    reference = model(**standard_tokenized).last_hidden_state
    flattened = model(**flattened_tokenized).last_hidden_state

    reference = reference[standard_tokenized["attention_mask"].bool()].unsqueeze(0)
cos = torch.nn.functional.cosine_similarity(reference[0], flattened[0], dim=-1)
print(cos.min().item(), cos.mean().item())
# 0.88525390625 0.9501953125
```

For reference, if I use `google-bert/bert-base-uncased`, I get `0.99951171875 1.0`.

### Expected behavior

I expect the previous script to return approximately `1.0 1.0`, just like with `bert-base-uncased`.

Having said that, `DataCollatorWithFlattening` knows nothing about the underlying model. It's the responsibility of the user to figure out whether the model supports it. Still, I think we should 1) update the `DataCollatorWithFlattening` docstring to note that some architectures require adding `padding_idx + 1` to the `position_ids` and to be careful, or we should 2) update `DataCollatorWithFlattening` to support a `position_ids_start` or something along those lines. Then, users can pass the correct value depending on the architecture.

I had an agent figure out which architectures were affected, and it's these:
`roberta`, `xlm_roberta`, `xlm_roberta_xl`, `camembert`, `roberta_prelayernorm`, `xmod`, `data2vec_text`, `longformer`, `luke`, `ibert`, `mpnet`, `markuplm`, `lilt`, `layoutlmv3`, `esm` (absolute mode, rotary ESM-2 is shift-invariant), and `pp_doclayout_v2`

The offset is uniformly exactly `padding_idx + 1`. All 16 store an int `padding_idx` on the embeddings module that owns the `position_embeddings` table. Notably `mpnet` hardcodes `padding_idx = 1` independent of `config.pad_token_id`, so a config-based correction would not be sufficient. Decoder-only models are unaffected (rotary positions are shift-invariant, and their sinusoidal `embed_positions` variants compute positions internally rather than consuming collator ids).

This includes very widely used checkpoints: every `xlm-roberta` based multilingual embedding model and `all-mpnet-base-v2` among them.

- Tom Aarsen

### PR DESCRIPTION
**Pull request by tomaarsen**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47525)

Fixes #47496

# What does this PR do?

## Pull Request overview
* Add a `position_ids_start` argument to `DataCollatorWithFlattening`
* Infer FlashAttention sequence boundaries from `position_ids == position_ids.min()` instead of `position_ids == 0`

## Details
As described in #47496, `DataCollatorWithFlattening` numbers positions from 0, but RoBERTa-like architectures (`roberta`, `xlm_roberta`, `camembert`, `mpnet`, `esm` with absolute position embeddings, and more) compute their position ids starting at `padding_idx + 1` (usually 2) when none are passed. Feeding these models the collator's 0-based `position_ids` silently shifts the position embeddings and degrades the outputs. This affects some very common checkpoints, e.g. `xlm-roberta`.

This PR adds a `position_ids_start` argument (default 0, existing behavior is unchanged) so users can match the numbering of their model, e.g. `position_ids_start=2` for XLM-R. The collator cannot reliably derive this from a config: `mpnet` for example hardcodes `padding_idx = 1` regardless of `config.pad_token_id`. The docstring now documents all arguments and warns about the pitfall.

The new argument alone turned out to be insufficient on the path where FlashAttention infers the varlen kwargs from `position_ids` at runtime (i.e. without `return_flash_attn_kwargs=True`, like the reproduction in the issue). `prepare_fa_kwargs_from_position_ids` detected sequence starts via `position_ids == 0`, which never matches offset positions, after which it crashes on an empty `cu_seq_lens.diff().max()`. I changed the detection to `position_ids == position_ids.min()`, mirroring `_is_packed_sequence` which already handles a non-zero start via `position_ids.min()`. Whenever a 0 position is present, this is behavior-identical, so existing models are unaffected.

I verified that with `position_ids_start=2` the collator output exactly matches `RobertaEmbeddings.create_position_ids_from_input_ids` on unpadded sequences, and that the RoBERTa embedding layer produces identical outputs for the flattened batch and for the individual sequences. The flattening collator tests now cover the new argument as well as the boundary inference on collator output.

Because of this `prepare_fa_kwargs_from_position_ids` patch, the PR got a bit bigger than I originally intended in https://github.com/huggingface/transformers/issues/47496#issuecomment-5058589235, but it does seem important.

## Code Agent Policy

- [ ] (First-time contributors only): I confirm that this PR description and code is not written by an LLM or code agent

The tests and docstrings were written by Fable and reviewed by myself. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@vasqu 

- Tom Aarsen

### CODE REVIEW

File: src/transformers/modeling_flash_attention_utils.py [hunk 1]
```diff
@@ -474,7 +474,9 @@ def prepare_fa_kwargs_from_position_ids(position_ids):
     tensor_kwargs = {"dtype": torch.int32, "device": position_ids.device}
 
     position_ids = position_ids.reshape(-1)
-    indices_q = (position_ids == 0).nonzero().view(-1)
+    # Packed sequences all restart from the same first position id, but it is not always 0
+    # (RoBERTa-like models start at padding_idx + 1)
+    indices_q = (position_ids == position_ids.min()).nonzero().view(-1)
```
Comment (vasqu) [inline src/transformers/modeling_flash_attention_utils.py:479]: Can we add an integration test for roberta that tests this case implicitly? It's ok to require FA or kernels

Comment (tomaarsen) [reply src/transformers/modeling_flash_attention_utils.py:479]: Yess, done in 9d1c8e8828d279916b498c05566940afcb17369b

File: tests/models/roberta/test_modeling_roberta.py [hunk 2]
```diff
@@ -505,6 +514,38 @@ def test_create_position_ids_from_inputs_embeds(self):
         self.assertEqual(position_ids.shape, expected_positions.shape)
         self.assertTrue(torch.all(torch.eq(position_ids, expected_positions)))
 
+    @require_flash_attn
+    @require_torch_accelerator
+    @pytest.mark.flash_attn_test
+    def test_flash_attn_padding_free_position_ids_start(self):
+        """Test that flattening a batch with position_ids_start=pad_token_id + 1 matches the padded batched run."""
+        config = self.model_tester.prepare_config_and_inputs()[0]
+        config.attn_implementation = "flash_attention_2"
```
Comment (vasqu) [inline tests/models/roberta/test_modeling_roberta.py]: let's use set_attn_implementation method instead of setting in the config

File: tests/models/roberta/test_modeling_roberta.py [hunk 3]
```diff
@@ -505,6 +514,38 @@ def test_create_position_ids_from_inputs_embeds(self):
...
+        model = RobertaModel(config).to(dtype=torch.bfloat16, device=torch_device).eval()
+
+        # Sample from [2, vocab_size) so no token collides with pad_token_id, which would shift
+        # the position ids the model computes internally in the batched reference run
+        seq_len_a, seq_len_b = 3, 4
+        input_ids = ids_tensor([2, seq_len_b], config.vocab_size - 2) + 2
+        attention_mask = torch.tensor([[1, 1, 1, 0], [1, 1, 1, 1]], device=torch_device)
+        input_ids[attention_mask == 0] = config.pad_token_id
```
Comment (vasqu) [inline tests/models/roberta/test_modeling_roberta.py]: Why not use the prepared inputs and just use a modified attention mask?

File: tests/models/roberta/test_modeling_roberta.py [hunk 4]
```diff
@@ -505,6 +514,38 @@ def test_create_position_ids_from_inputs_embeds(self):
...
+        self.assertEqual(packed["position_ids"][0].tolist(), [2, 3, 4, 2, 3, 4, 5])
+        # Without attention_mask and cu_seq_lens, FlashAttention infers the boundaries from position_ids
+        with torch.no_grad():
+            flattened = model(
+                input_ids=packed["input_ids"].to(torch_device), position_ids=packed["position_ids"].to(torch_device)
+            ).last_hidden_state
+
+        # bf16 noise between the padded and packed kernel shapes reaches ~3e-2, wrong position ids differ by >1.9
+        torch.testing.assert_close(batched[0, :seq_len_a], flattened[0, :seq_len_a], atol=1e-1, rtol=1e-1)
+        torch.testing.assert_close(batched[1, :seq_len_b], flattened[0, seq_len_a:], atol=1e-1, rtol=1e-1)
```
Comment (vasqu) [inline tests/models/roberta/test_modeling_roberta.py]: I'd rather use fp16 then + a lower tol, this seems quite high no?

**Review by vasqu — APPROVED** [review]
In general yes, just some comments for the test design

**Comment by tomaarsen** [comment]
I moved to using `input_ids` from `prepare_config_and_inputs`, though this required some more minor changes in the rest of the test. I also moved to fp16, which did help lower the tol's to 1e-3. I also tested with lower tolerances and that was fine as well, but I figured I'd keep it reasonable given that the CI / other users will have different hardware.

- Tom Aarsen

**Review by vasqu — APPROVED** [review]
Looks good now, thanks!
